connection-cache/src/connection_cache.rs

```diff
@@ -156,6 +197,60 @@ where
         }
     }
 
+    fn create_connection_internal(
+        config: Arc<C>,
+        connection_manager: Arc<M>,
+        map: &mut std::sync::RwLockWriteGuard<'_, IndexMap<SocketAddr, P>>,
+        addr: &SocketAddr,
+        connection_pool_size: usize,
+        async_connection_sender: Option<Sender<(usize, SocketAddr)>>,
+    ) -> (bool, u64, u64) {
+        // evict a connection if the cache is reaching upper bounds
+        let mut num_evictions = 0;
+        let mut get_connection_cache_eviction_measure =
+            Measure::start("get_connection_cache_eviction_measure");
+        let existing_index = map.get_index_of(addr);
+        while map.len() >= MAX_CONNECTIONS {
+            let mut rng = thread_rng();
+            let n = rng.gen_range(0..MAX_CONNECTIONS);
+            if let Some(index) = existing_index {
+                if n == index {
+                    continue;
+                }
+            }
+            map.swap_remove_index(n);
+            num_evictions += 1;
+        }
+        get_connection_cache_eviction_measure.stop();
+
+        let mut hit_cache = false;
+        map.entry(*addr)
+            .and_modify(|pool| {
+                if pool.need_new_connection(connection_pool_size).0 {
+                    pool.add_connection(&config, addr);
+                    async_connection_sender.map(|sender| {
+                        debug!(
+                            "Sending async connection creation {} for {addr}",
+                            pool.num_connections() - 1
+                        );
+                        sender.send((pool.num_connections() - 1, *addr)).unwrap();
+                    });
```

Comment:
Rather than hardcoding num_connections() - 1, how about returning the index of the created item in add_connection() and then using that? This way you're always sure warm up the correct connection. With the lock, I don't think it's possible to use the wrong connection, but at least this way you're always sure